### PR TITLE
[l4] Fix the I2C on STM32L4

### LINKED ISSUE #255: [I2C / L4] Restart Condition not generated and nextTransaction=Write not supported.
These issues were discovered when using the SSD1306 display with the hardware I2C. Transactions with more than 255 bytes need different sequence in STM32L4.
PR is in development.

### PR DESCRIPTION
**Pull request by strongly-typed**
Finally, for the 2017q4 release. Shall fix #255.

Restart condition is now correctly generated.

screen shot 2017-10-12 at 00 43 35

Integration tested with SSD1306 display (write > 255 bytes).

Test coverage is still a bit light, so expect surprises.
```
➜  sr_macosx ./bin/sigrok-cli --driver fx2lafw --config samplerate=1m --samples 64k --channels D4=sda,D5=scl --triggers sda=f --wait-trigger -P i2c:sda=sda:scl=scl | grep -v "i2c-1: 0" | grep -v "i2c-1: 1" | sed s/Stop/Stop\\\n==========/g
i2c-1: Start
i2c-1: Write
i2c-1: Address write: 3C
i2c-1: ACK
i2c-1: Stop
==========
i2c-1: Start
i2c-1: Write
i2c-1: Address write: 3C
i2c-1: ACK
i2c-1: Data write: 01
i2c-1: ACK
i2c-1: Stop
==========
i2c-1: Start
i2c-1: Write
i2c-1: Address write: 3C
i2c-1: ACK
i2c-1: Data write: 01
i2c-1: ACK
i2c-1: Data write: 02
i2c-1: ACK
i2c-1: Stop
==========
i2c-1: Start
i2c-1: Read
i2c-1: Address read: 3C
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: NACK
i2c-1: Stop
==========
i2c-1: Start
i2c-1: Write
i2c-1: Address write: 3C
i2c-1: ACK
i2c-1: Data write: 01
i2c-1: ACK
i2c-1: Start repeat
i2c-1: Read
i2c-1: Address read: 3C
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: NACK
i2c-1: Stop
==========
i2c-1: Start
i2c-1: Write
i2c-1: Address write: 3C
i2c-1: ACK
i2c-1: Data write: 01
i2c-1: ACK
i2c-1: Data write: 02
i2c-1: ACK
i2c-1: Start repeat
i2c-1: Read
i2c-1: Address read: 3C
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: ACK
i2c-1: Data read: 03
i2c-1: NACK
i2c-1: Stop
==========```

### CODE REVIEW

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in [hunk 1]
```diff
@@ -76,6 +76,54 @@ namespace
...
+			// RELOAD if the operation continues
+			reload = true;
+		} else if ((nextOperation == xpcc::I2c::Operation::Stop) /* and (writing.length <= 0xff) */) {
+			// AUTOEND has no effect if RELOAD is set.
+			autoend = true;
+		}
+
+		DEBUG_STREAM("autoend=" << autoend << ", reload=" << reload << ", nbytes=" << nbytes);
+
+		I2C{{ id }}->CR2 = (autoend ? I2C_CR2_AUTOEND : 0 ) |\
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: `|\`? Shouldn't be necessary for code, only for defines.

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in [hunk 2]
```diff
@@ -124,6 +136,58 @@ namespace
...
+			// RELOAD if the operation continues
+			reload = true;
+		} else if ((nextOperation == xpcc::I2c::Operation::Stop)) {
+			// AUTOEND has no effect if RELOAD is set.
+			autoend = true;
+		}
+
+		DEBUG_STREAM("autoend=" << autoend << ", reload=" << reload << ", nbytes=" << nbytes);
+
+		I2C{{ id }}->CR2 = (autoend ? I2C_CR2_AUTOEND : 0) |\
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: Same here with `|\`.

File: src/xpcc/architecture/platform/board/nucleo_l476rg/nucleo_l476rg.cpp [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 // Create an IODeviceWrapper around the Uart Peripheral we want to use
 Board::LoggerDevice loggerDevice;
+xpcc::IOStream stream(loggerDevice);
```
Comment (salkinium) [inline src/xpcc/architecture/platform/board/nucleo_l476rg/nucleo_l476rg.cpp]: Do we want to expose this at a board level? Maybe move to the `main.cpp` of the example. In general debugging/tracing could be improved in xpcc, but it's a hard problem, and Uart works reasonably okish.

**Review by salkinium — CHANGES_REQUESTED** [review]
Small cosmetic changes. Otherwise 👍.

File: src/xpcc/architecture/platform/board/nucleo_l476rg/nucleo_l476rg.cpp (hunk 3 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/board/nucleo_l476rg/nucleo_l476rg.cpp]: > move to the main.cpp of the example

Must be globally available by the driver. I do not like `extern` or whatever.
Maybe it is just left to the user to add the `stream` when necessary for debugging. Debugging I2C is not for the faint-hearted, btw.

File: src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in (hunk 2 above)
Comment (strongly-typed) [reply src/xpcc/architecture/platform/driver/i2c/stm32l4/i2c_master.cpp.in]: More is always better, you know! Just fixed.

**Review by salkinium — DISMISSED** [review]
🚢🕴 Please upstream with a custom message merge commit.

**Pull request merged**